### ISSUE TITLE
New buzzwords A Primer on Nvidia-Docker  Where Containers Meet GPUs

### ISSUE THREAD

**Issue by sudocurse** [sudocurse/reading-list#7804]
A Primer on Nvidia-Docker &mdash; Where Containers Meet GPUs<br>
http://ift.tt/2FvlFKb<br>
<img width="640" height="457" src="https://storage.googleapis.com/cdn.thenewstack.io/media/2018/03/0fc04f33-barracuda-329561_640.jpg" class="webfeedsFeaturedVisual wp-post-image wp-stateless-item" style="display: block; margin: auto; margin-bottom: 20px;" data-image-size="large" data-stateless-media-bucket="cdn.thenewstack.io" data-stateless-media-name="media/2018/03/0fc04f33-barracuda-329561_640.jpg">
<p>GPUs are critical for training deep learning models and neural networks. Though it may not be needed for simple models based on linear regression and logistic regression, complex models designed around convolutional neural networks (CNNs) and recurrent neural networks heavily rely on GPUs. Especially computer vision-related models based on frameworks such as Caffe2 and TensorFlow have a dependency on GPU.</p>
<p>In supervised machine learning, a set of features and labels are used to train a model. Deep learning algorithms don&rsquo;t even need explicit features to evolve trained models. They pretty much &ldquo;learn&rdquo; from existing datasets designated for training, testing, and evaluation.</p>
<p>Neural networks perform complex computation on tens of thousands of matrices before the final model is evolved. When an image is fed to a CNN, it gets translated into a matrix of real numbers. Depending on the density and size of the image, multiple such matrices are generated by the neural network. These matrices are added and multiplied with other matrices during the forward propagation and backward propagation till appropriate weights are derived.</p>
<p>A trained model can be run on CPUs for inference. Since inference is not as intense as training, GPUs are strictly optional when running models for inference.</p>
<p>CPUs are not designed to deal with such a rapid rate of computation. While they are faster for performing regular number crunching, CPUs are not meant for parallelizing mathematical operations. That&rsquo;s where GPU plays a crucial role. They may not have the horsepower of CPUs, but they can perform massively parallelized calculations.</p>
<h2>Nvidia CUDA and cuDNN</h2>
<p>Traditional programs cannot access GPUs directly. They need a special parallel programming interface to move computations to GPU. Nvidia, the most popular graphics card manufacturer, has created <a href="http://developer.download.nvidia.com/compute/cuda/1.0/NVIDIA_CUDA_Programming_Guide_1.0.pdf" class="ext-link" rel="external nofollow">Compute Unified Device Architecture</a> (CUDA), as a parallel computing platform and programming model for general computing on GPUs. With CUDA, developers will be able to dramatically speed up computing applications by harnessing the power of GPUs.</p>
<p>In GPU-enabled applications, the sequential part of the workload continues to run on the CPU &mdash; which is optimized for single-threaded performance &mdash; while the parallelized compute intensive part of the application is offloaded to run on thousands of GPU cores in parallel. To integrate CUDA, developers program in popular languages such as C, C++, Fortran, Python and MATLAB by expressing parallelism through extensions in the form of a few basic keywords.</p>
<p>Deep learning networks and neural networks heavily rely on CUDA for parallel computation. Nvidia has also developed a specialized library called <a href="https://developer.nvidia.com/cudnn" class="ext-link" rel="external nofollow">CUDA Deep Neural Network library</a> (cuDNN), which is a GPU-accelerated library of primitives for deep neural networks. cuDNN provides highly tuned implementations for standard routines such as forward and backward convolution, pooling, normalization, and activation layers.</p>
<p>Deep learning researchers and framework developers worldwide rely on cuDNN for high-performance GPU acceleration. It allows them to focus on training neural networks and developing software applications rather than spending time on low-level GPU performance tuning. cuDNN accelerates widely used deep learning frameworks, including Caffe2, MATLAB, Microsoft Cognitive Toolkit, TensorFlow, Theano, and PyTorch.</p>
<h2>Nvidia-Docker &mdash; Bridging the Gap Between Containers and GPU</h2>
<p>In 2016, Nvidia created a runtime for Docker called <a href="https://github.com/NVIDIA/nvidia-docker" class="ext-link" rel="external nofollow">Nvidia-Docker</a>. The goal of this open source the project was to bring the ease and agility of containers to CUDA programming model.</p>
<p>Since Docker didn&rsquo;t support GPUs natively, this project instantly became a hit with the CUDA community. Nvidia-Docker is basically a wrapper around the docker CLI that transparently provisions a container with the necessary dependencies to execute code on the GPU. It is only necessary when using Nvidia-Docker run to execute a container that uses GPUs.</p>
<p>You can run Nvidia-Docker on Linux machines that have a GPU along with the required drivers installed. Refer to the <a href="https://github.com/NVIDIA/nvidia-docker/blob/master/README.md" class="ext-link" rel="external nofollow">readme</a> on Nvidia-Docker&rsquo;s <a href="https://github.com/NVIDIA/nvidia-docker" class="ext-link" rel="external nofollow">Github</a> repository for details on the prerequisites.</p>
<p><img class="aligncenter size-full wp-image-4313223" src="https://storage.googleapis.com/cdn.thenewstack.io/media/2018/03/d9c54f25-nvidia-docker.png" width="522" height="369" data-id="4313223"></p>
<p>Below are the commands I used to install and verify Nvidia-Docker on an Ubuntu 16.04 machine powered by a Nvidia Quadro P4000 GPU.</p>
<pre>
$ curl -s -L https://nvidia.github.io/nvidia-docker/gpgkey |  sudo apt-key add -

$ curl -s -L https://nvidia.github.io/nvidia-docker/ubuntu16.04/amd64/nvidia-docker.list | sudo tee /etc/apt/sources.list.d/nvidia-docker.list

$ sudo apt-get update

$ sudo apt-get install -y nvidia-docker

$ sudo pkill -SIGHUP dockerd # Restart Docker Engine

$ sudo nvidia-docker run --rm nvidia/cuda nvidia-smi
</pre>
<p><img class="aligncenter size-full wp-image-4313225" src="https://storage.googleapis.com/cdn.thenewstack.io/media/2018/03/689e2ec1-nvidia-smi.jpg" width="1372" height="690" data-id="4313225"></p>
<p>The nvidia-smi command runs the systems management interface (SMI) to confirm that the Docker container is able to access the GPU. Behind the scenes, SMI talks to the Nvidia driver to talk to the GPU.</p>
<p>We can also verify that CUDA is installed by running the below command.</p>
<pre>
$ sudo nvidia-docker run --rm nvidia/cuda nvcc -V
</pre>
<p><img class="aligncenter size-full wp-image-4313226" src="https://storage.googleapis.com/cdn.thenewstack.io/media/2018/03/666b428c-nvcc.jpg" width="913" height="272" data-id="4313226"></p>
<p>Data scientists and developers can use this Docker image as the base for running popular neural network frameworks such as TensorFlow, Apache MXNext, Microsoft CNTK, and Caffe2.</p>
<h2>How about Kubernetes?</h2>
<p><a href="https://cloud.google.com/kubernetes-engine/" class="ext-link" rel="external nofollow">Google Kubernetes Engine</a> (GKE) has added beta support for GPUs. It is one of the first managed Kubernetes services to add support for Nvidia Tesla K80 and Tesla P100 GPUs. This addition opens up interesting avenues for developers and data scientists. We can now train deep learning models and neural networks in Kubernetes. It&rsquo;s not just machine learning, any parallelizable workload that needs access to GPUs such as video rendering or game development can take advantage of this feature.</p>
<p>As a deep learning enthusiast with a passion for containers, I was very excited to see GPU support for Kubernetes. I ported a few Caffe models to GKE to test the waters. The overall experience has been pretty smooth for a beta release.</p>
<p>In the upcoming <a href="https://www.thenewstack.io/tag/Kubernetes-GPU-series" class="local-link">multipart tutorial</a>, I will walk you through all the steps required to create a GPU-based Kubernetes cluster, training a <a href="https://www.thenewstack.io/tag/convolutional-neural-network" class="local-link">convolutional neural network</a> (CNN), and running the training model for inference in the development machine that doesn&rsquo;t have a GPU. Though the tutorial doesn&rsquo;t delve into the concepts of deep learning and CNNs, it does give you everything you need to jumpstart model training on Kubernetes.</p>
<p>Stay tuned!</p>
<p class="attribution"><a href="http://bit.ly/2x5VLZD" class="ext-link" rel="external nofollow">Google</a> is a sponsor of The New Stack.</p>
<p>Feature image <a href="https://pixabay.com/en/barracuda-fish-cayman-islands-scuba-329561/" class="ext-link" rel="external nofollow">via</a> Pixabay.</p>
<p>The post <a rel="nofollow" href="https://thenewstack.io/primer-nvidia-docker-containers-meet-gpus/">A Primer on Nvidia-Docker &mdash; Where Containers Meet GPUs</a> appeared first on <a rel="nofollow" href="https://thenewstack.io">The New Stack</a>.</p>
<br>
<br>
via Blog &ndash; The New Stack https://thenewstack.io<br>
March 09, 2018 at 11:33AM
